packages/flutter/lib/src/widgets/actions.dart

```diff
@@ -276,88 +520,169 @@ class Actions extends InheritedWidget {
-      // Will only get here if nullOk is true.
-      return false;
-    }
-
-    // Invoke the action we found using the dispatcher from the Actions Element
-    // we found, using the given focus node.
-    return _findDispatcher(actionsElement).invokeAction(action, intent, focusNode: focusNode);
+    // Invoke the action we found using the relevant dispatcher from the Actions
+    // Element we found.
+    return actionElement != null ? _findDispatcher(actionElement).invokeAction(action, intent, context) != null : null;
   }
 
   @override
-  bool updateShouldNotify(Actions oldWidget) {
-    return oldWidget.dispatcher != dispatcher || !mapEquals<LocalKey, ActionFactory>(oldWidget.actions, actions);
-  }
+  State<Actions> createState() => _ActionsState();
 
   @override
   void debugFillProperties(DiagnosticPropertiesBuilder properties) {
     super.debugFillProperties(properties);
     properties.add(DiagnosticsProperty<ActionDispatcher>('dispatcher', dispatcher));
-    properties.add(DiagnosticsProperty<Map<LocalKey, ActionFactory>>('actions', actions));
+    properties.add(DiagnosticsProperty<Map<Type, Action<Intent>>>('actions', actions));
+  }
+}
+
+class _ActionsState extends State<Actions> {
+  // Keeps the last-known enabled state of each action in the action map.
+  Map<Action<Intent>, bool> enabledState = <Action<Intent>, bool>{};
+  // Used to tell the market to rebuild when the enabled state of an action in
+  // the map changes.
+  LocalKey rebuildKey = UniqueKey();
+
+  @override
+  void initState() {
+    super.initState();
+    _updateActionListeners();
+  }
+
+  void _handleActionChanged(Action<Intent> action) {
+    assert(enabledState.containsKey(action));
+    final bool actionEnabled = action.enabled;
+    if (enabledState[action] == null || enabledState[action] != actionEnabled) {
+      setState(() {
+        enabledState[action] = actionEnabled;
+        // Generate a new key so that the marker notifies dependents.
+        rebuildKey = UniqueKey();
+      });
+    }
+  }
+
+  void _updateActionListeners() {
+    final Map<Action<Intent>, bool> newState = <Action<Intent>, bool>{};
+    final Set<Action<Intent>> foundActions = <Action<Intent>>{};
+    for (final Action<Intent> action in widget.actions.values) {
+      if (enabledState.containsKey(action)) {
+        // Already subscribed to this action, just copy over the current enabled state.
+        newState[action] = enabledState[action];
+        foundActions.add(action);
+      } else {
+        // New action to subscribe to.
+        // Don't set the new state to action.enabled, since that can cause
+        // problems when the enabled accessor looks up other widgets (which may
+        // have already been removed from the tree).
+        newState[action] = null;
+        action.addActionListener(_handleActionChanged);
+      }
+    }
+    for (final Action<Intent> action in enabledState.keys) {
+      if (!foundActions.contains(action)) {
```

Comment:
Could you just remove found actions from enabledState in the for loop above to avoid this extra check? (and also the extra foundAction set could be removed then).

Reply:
lines 637/638 effectively iterate over all actions in enabledState that were not added to foundActions. So in line 625 could you just do `enabledActions.removeKey(action)` instead of `foundActions.add(action)`? At the end of the first loop, `enabledActions` would only contain the actions that were not preserved, and in the second loop you could just iterate over enabledState.keys and remove their listeners unconditionally. That's what I meant. But if that sounds more complicated, I am also ok with keeping this as-is.

This of course only works if widget.actions.values doesn't include the same action twice....

Reply:
Actually, your current implementation is a little more resilient to duplicated actions, so maybe just leave this as-is.